### DISCUSSION TITLE
Realtime processing: output display?

### DISCUSSION THREAD

**Discussion by jmfriedt** [aewallin/allantools#153] [General]
May I inquire about the realtime processing API? I fail to identify how to recover the {tau,mdev} values after new samples have been pushed.
I am confused by the example ``realtime_benchmark.py`` with
```Python
 x = numpy.random.randn(pow(10, decades+2))
 for x in range(pow(10, decades)):
        dev_rt.add_phase(x)
```
where the random data sequence seems to be overwritten by the decade definition. When trying to add calls to dev_rt.taus() I get realistic integration delays but calling dev_rt.devs() always returns an array of 0s with different length than taus.

Can enyone enlighten on the usage of the realtime API for recovering and displaying the ongoing Allan deviation calculation as new samples are pushed?
Thanks.

**Comment by jmfriedt** [comment]
I remain suspicious of the benchmark, especially after successfully integrating the real time processing in GNU Radio at https://github.com/jmfriedt/gr-allan. For reference, https://github.com/jmfriedt/gr-allan/blob/main/python/allan/allantxt.py should provide a simple and compact example for accessing the results of the real time computation without the overhead of the full graphical display demonstrated at https://github.com/jmfriedt/gr-allan/blob/main/python/allan/allan.py

Thank you for the excellent library.

**Comment by aewallin** [comment]
The tests have a more realistic use-case, maybe they can be of help:
https://github.com/aewallin/allantools/tree/master/tests/realtime

I just created a short example as a github gist here:
https://gist.github.com/aewallin/8caf2dbdeba4e0008ae697c6aeafffa5

we are also using Ettus N210 and B200 sdrs as frequency counters, so testing out a realtime ADEV-display for gnuradio would be very interesting - please keep us updated on progress if/when you have time!

Anders

**Comment by jmfriedt** [comment]
I'll give the B210 a test on my own as well as phase detector, but the https://github.com/jmfriedt/gr-allan repository sounds quite functional now (GNU Radio 3.10) and ready to be fed with real phase samples rather than simulated datasets as was done so far.

Edit:
I have added in https://github.com/jmfriedt/gr-allan/tree/main/examples an example of a measurement with a B210 SDR connected to a TTi TGR2053 free running synthesizer (no hydrogen maser control) at 600 MHz. The output seems more or less consistent with Fig 2 of https://www.ursi.fi/2019/Papers/Hanhijarvi.pdf if I did not make a mistake in the phase to time conversion (tau from 1e-2 to 1e2 seconds):

![2023-06-12-164518_2704x1050_scrot](https://github.com/aewallin/allantools/assets/10595447/96d6b43d-aea3-4fc3-afeb-8cc3440b8cef)

Thanks
